File: bonus_content/tik_tok_effect_for_children/tik_tok_effect_src/Emoji Puzzles 4 Kids/Graph/ScriptTemplate/Wait/CGWait.lua
local CGWait = CGWait or {}
CGWait.__index = CGWait

function CGWait.new()
    local self = setmetatable({}, CGWait)
    self.inputs = {}
    self.outputs = {}
    self.nexts = {}
    self.timeUsed = 0
    self.enable = false
    return self
end

function CGWait:setNext(index, func)
    self.nexts[index] = func
end

function CGWait:setInput(index, func)
    self.inputs[index] = func
end

function CGWait:getOutput(index)
    return self.inputs[index]()
end

function CGWait:execute()
    self.enable = true
end

function CGWait:update(sys, deltaTime)
    if not self.enable then
        return
    end

    self.timeUsed = deltaTime + self.timeUsed
    if self.timeUsed < self.inputs[1]() then
        return
    end
    if self.nexts[0] then
        self.timeUsed = 0
        self.enable = false
        self.nexts[0]()
    end
end

return CGWait

File: bonus_content/tik_tok_effect_for_children/tik_tok_effect_src/Emoji Puzzles 4 Kids/Graph/ScriptTemplate/Wait/CGWait.js
/**
 * @file CGWait.js
 * @author liujiacheng
 * @date 2021/8/20
 * @brief CGWait.js
 * @copyright Copyright (c) 2021, ByteDance Inc, All Rights Reserved
 */

const {BaseNode} = require('./BaseNode');
const Amaz = effect.Amaz;

class CGWait extends BaseNode {
  constructor() {
    super();
    this.timeUsed = 0;
    this.enable = false;
  }

  execute(index) {
    this.enable = true;
  }

  onUpdate(sys, dt) {
    if (!this.enable) {
      return;
    }

    this.timeUsed = this.timeUsed + dt;
    if (this.timeUsed < this.inputs[1]()) {
      return;
    }

    if (this.nexts[0]) {
      this.timeUsed = 0;
      this.enable = false;
      this.nexts[0]();
    }
  }
}

exports.CGWait = CGWait;
